source/server/listener_manager_impl.cc

```diff
@@ -694,20 +706,49 @@ void ListenerManagerImpl::onListenerWarmed(ListenerImpl& listener) {
   updateWarmingActiveGauges();
 }
 
-void ListenerManagerImpl::drainFilterChains(ListenerImplPtr&& listener) {
+void ListenerManagerImpl::inPlaceFilterChainUpdate(ListenerImpl& listener) {
+  auto existing_active_listener = getListenerByName(active_listeners_, listener.name());
+  auto existing_warming_listener = getListenerByName(warming_listeners_, listener.name());
+  ASSERT(existing_warming_listener != warming_listeners_.end());
+  ASSERT(*existing_warming_listener != nullptr);
+
+  (*existing_warming_listener)->debugLog("execute in place filter chain update");
+  if (existing_active_listener != active_listeners_.end()) {
+    ASSERT(*existing_active_listener != nullptr);
+    // The warmed listener should be added first so that the worker will accept new connections
+    // when it stops listening on the old listener.
+    for (const auto& worker : workers_) {
+      addListenerToWorker(*worker, listener.listenerTag(), listener, nullptr);
+    }
+    // Finish active_listeners_ transformation before calling `drainListener` as it depends on their
+    // state.
```

Comment:
Please add more comments here. I'm not totally sure what the sequence of events here is in terms of how listeners cycle through their lifetime.

Reply:
This comment is copied from onListenerWarmed. OK in comment I need to replace `drainListener` by "drainFilterChains`.

If you are not referring to the copy paste error: The `drainListener` and `drainFilterChains` schedules future tasks and captures the `iterator`. 

Suppose we call drainXXX before the transfer from warm to active, bellow would happen

1. call drainXXX(), schedule the callback in 3, capture iter
2. transfer iter from warm to active
3. callback now the iter of "previous active" is now referring to the "previous warm"

If we swap 1) and 2), `iter` will always refer the same ListenerImpl.